src/architecture/merkledb.md

```diff
@@ -0,0 +1,207 @@
+The following operations are supported:
+
+- Get and set a list item by index
+- Append an item to the list
+- Pop or poll the last item from the list
+- Get the list length
+- Check if the list is empty
+- Iterate over index-item pairs ordered by index
+- Insert a sequence of items from an iterator
+- Truncate the list to the specified length
+- Clear the list (i.e., remove all stored items from the list).
+
+`ListIndex` does not support inserting items in the middle of the
+list or removing items by index
+(although it is still possible to implement these operations manually).
+
+!!! summary "Implementation Details"
+    To support proper iteration, `ListIndex` saves its items with 8-byte
+    unsigned indices as keys. The indices are serialized in the big-endian form.
+
+### SparseListIndex
+
+[`SparseListIndex`][sparse-list-index] represents a `ListIndex` that may
+contain "gaps". It provides the possibility to delete items not only from
+the end of the list, but from any part thereof. Such deletions do not break
+the order of the indices inside the list.
+
+The remaining functionality of the `SparseListIndex` is the same as for
+[`ListIndex`](#ListIndex).
+
+### ValueSetIndex
+
+[`ValueSetIndex`][value-set-index] represents a hash set.
+The following operations are implemented:
+
+- Add and remove set items
+- Check if an item is already present using the item itself or its hash
+- Iterate over stored items in the lexicographic order of their hashes
+- Iterate over hashes of items in the lexicographic order
+- Clear the set (i.e., remove all items).
+
+The hash used in `ValueSetIndex` is calculated using the `object_hash()` method
+of the [`ObjectHash` trait][object-hash].
+
+!!! summary "Implementation Details"
+    Internally, `ValueSetIndex` uses hashes of items as keys,
+    and items themselves as corresponding values.
+
+### KeySetIndex
+
+[`KeySetIndex`][key-set-index] represents a set.
+The following procedures are implemented:
+
+- Add and remove set items
+- Check if a specific item is in the set
+- Iterate over items in the lexicographic order
+- Clear the set (i.e., remove all stored items).
+
+!!! summary "Implementation Details"
+    Internally, the item is used as a key, and its value is always empty.
+
+#### KeySetIndex vs ValueSetIndex
+
+While `ValueSetIndex` uses a hash as a key, `KeySetIndex` puts an entire binary
+serialization of an item into the key.
+
+- `KeySetIndex` does not have an additional overhead on hashing
+  set items.
+- `KeySetIndex` should not be used when set items are relatively big;
+  only small items should be stored in it (such as integers, small strings,
+  small
+  tuples). On the other hand, `ValueSetIndex` handles storing big and complex
+  items more easily.
+- `KeySetIndex` introduces a lexicographical order over stored
+  items. In `ValueSetIndex` items are ordered according to their hash
+  function properties.
+
+### Entry
+
+`Entry` represents a set that contains only one item.
```

Comment:
I wouldn't call it a set as it is more akin to an `Optional` — a container that may contain a single item or be empty (= no item).